Description: Munin attempts to upgade values in RRD files created by older
 versions of munin, and rename those files, but on first startup, the file
 /var/lib/munin/datafile does not exist, and $oldconfig does not contain the
 expected values.
 .
 The patch attemts to fix this by adding checks for the presence of the old
 datafile, by adding checks for old RRD file, and by adding checks for the
 content of the config hash when used in Munin.pm
 .
 This patch will not apply to munin upstream after 1.3.4, due to an extensive
 rewrite / refactoring.
Bug-Debian: http://bugs.debian.org/400703
Author: Stig Sandbeck Mathisen <ssm@debian.org>
--- a/server/Munin.pm.in
+++ b/server/Munin.pm.in
@@ -572,9 +572,29 @@
 
 sub munin_get_filename {
 	my ($config,$domain,$node,$service,$field) = @_;
+    my ($graphtype, $suffix);
 
-	return ($config->{'dbdir'} . "/$domain/$node-$service-$field-" . lc substr (($config->{domain}->{$domain}->{node}->{$node}->{client}->{$service}->{$field.".type"}||"GAUGE"), 0,1). ".rrd");
+    if ($config->{'dbdir'})
+    {
+	if ($config->{domain}->{$domain}->{node}->{$node}->{client}->{$service}->{$field . ".type"})
+	{
+	    $graphtype = $config->{domain}->{$domain}->{node}->{$node}->{client}->{$service}->{$field . ".type"};
+	}
+	else
+	{
+	    # default graph type
+	    $graphtype = "GAUGE";
+	}
+
+	# Suffix for the filename is lowercase first letter of the RRD graph type
+	$suffix = lc(substr($graphtype, 0, 1));
 
+	return ($config->{'dbdir'} . "/$domain/$node-$service-$field-$suffix.rrd");
+    }
+    else
+    {
+        return 0;
+    }
 }
 
 sub munin_get_bool
--- a/server/munin-update.in
+++ b/server/munin-update.in
@@ -295,7 +295,7 @@
 my $overwrite = &munin_readconfig($conffile);
 $config = &munin_overwrite($config,$overwrite);
 
-&compare_configs ($oldconfig, $config);
+&compare_configs ($oldconfig, $config) if $oldconfig;
 
 if (&munin_getlock("$config->{rundir}/munin-datafile.lock")) {
     &munin_writeconfig("$config->{dbdir}/datafile",$config);
@@ -407,11 +407,14 @@
     my $ofile  = &munin_get_filename ($oconf, $domain, $host, $serv, $field);
     my $nfile  = &munin_get_filename ($nconf, $domain, $host, $serv, $field);
 
-    logger ("INFO: Changing type of $domain -> $host -> $serv -> $field to " . (defined $val?$val:"GAUGE") . ".\n");
-    RRDs::tune ($ofile, "-d", "42:".(defined $val?$val:"GAUGE"));
-    unless (rename ($ofile, $nfile))
+    if ( $ofile && -e $ofile )
     {
-	logger ("ERROR: Could not rename file: $!\n");
+      logger ("INFO: Changing type of $domain -> $host -> $serv -> $field to " . (defined $val?$val:"GAUGE") . ".\n");
+      RRDs::tune ($ofile, "-d", "42:".(defined $val?$val:"GAUGE"));
+      unless (rename ($ofile, $nfile))
+	{
+	    logger ("ERROR: Could not rename file: $!\n");
+	}
     }
 }
 
